cpp/src/arrow/compute/kernels/scalar_temporal_test.cc

```diff
@@ -0,0 +1,239 @@
+  auto time_points_nano = ArrayFromJSON(timestamp(TimeUnit::NANO), json_nano);
+
+  auto year = ArrayFromJSON(int64(), "[1970, 2000, 1900, 2033]");
+  auto month = ArrayFromJSON(int64(), "[1, 2, 1, 5]");
+  auto day = ArrayFromJSON(int64(), "[1, 29, 1, 18]");
+  auto day_of_year = ArrayFromJSON(int64(), "[1, 60, 1, 138]");
+  auto week = ArrayFromJSON(int64(), "[1, 9, 1, 20]");
+  auto quarter = ArrayFromJSON(int64(), "[1, 1, 1, 2]");
+  auto day_of_week = ArrayFromJSON(int64(), "[4, 2, 1, 3]");
+  auto hour = ArrayFromJSON(int64(), "[0, 23, 1, 3]");
+  auto minute = ArrayFromJSON(int64(), "[0, 23, 59, 33]");
+  auto second = ArrayFromJSON(int64(), "[59, 23, 20, 20]");
+  auto millisecond = ArrayFromJSON(int64(), "[0, 0, 0, 0]");
+  auto microsecond = ArrayFromJSON(int64(), "[0, 0, 0, 0]");
+  auto nanosecond = ArrayFromJSON(int64(), "[0, 0, 0, 0]");
+
+  for (auto time_points :
+       {time_points_second, time_points_milli, time_points_micro, time_points_nano}) {
+    CheckScalarUnary("year", time_points, year);
+    CheckScalarUnary("month", time_points, month);
+    CheckScalarUnary("day", time_points, day);
+    CheckScalarUnary("day_of_year", time_points, day_of_year);
+    CheckScalarUnary("week", time_points, week);
+    CheckScalarUnary("quarter", time_points, quarter);
+    CheckScalarUnary("day_of_week", time_points, day_of_week);
+    CheckScalarUnary("hour", time_points, hour);
+    CheckScalarUnary("minute", time_points, minute);
+    CheckScalarUnary("second", time_points, second);
+    CheckScalarUnary("millisecond", time_points, millisecond);
+    CheckScalarUnary("microsecond", time_points, microsecond);
+    CheckScalarUnary("nanosecond", time_points, nanosecond);
+  }
+
+  std::string in = "[123, 999, 1, 31231000]";
+  auto out = ArrayFromJSON(int64(), "[123, 999, 1, 0]");
+
+  auto tp_milli = ArrayFromJSON(timestamp(TimeUnit::MILLI), in);
+  auto tp_milli_zoned = ArrayFromJSON(timestamp(TimeUnit::MILLI, "Etc/GMT+2"), in);
+  CheckScalarUnary("millisecond", tp_milli, out);
+  CheckScalarUnary("millisecond", tp_milli, out);
+
+  auto tp_micro = ArrayFromJSON(timestamp(TimeUnit::MICRO), in);
+  auto tp_micro_zoned = ArrayFromJSON(timestamp(TimeUnit::MICRO, "Etc/GMT+2"), in);
+  CheckScalarUnary("microsecond", tp_micro, out);
+  CheckScalarUnary("microsecond", tp_micro_zoned, out);
+
+  auto tp_nano = ArrayFromJSON(timestamp(TimeUnit::NANO), in);
+  auto tp_nano_zoned = ArrayFromJSON(timestamp(TimeUnit::NANO, "Etc/GMT+2"), in);
+  CheckScalarUnary("nanosecond", tp_nano, out);
+  CheckScalarUnary("nanosecond", tp_nano_zoned, out);
+}
+
+TEST(ScalarTemporalTest, TestSimpleZonedTemporalComponentExtraction) {
+  const char* json =
+      R"(["1970-01-01T00:00:59","2000-02-29T23:23:23",
+          "3989-07-14T18:04:01","1900-01-01T01:59:20","2033-05-18T03:33:20"])";
+  auto time_points = ArrayFromJSON(timestamp(TimeUnit::SECOND, "Etc/GMT+2"), json);
+
+  auto year = ArrayFromJSON(int64(), "[1969, 2000, 3989, 1899, 2033]");
```

Comment:
Hmm... can you explain the reasoning here? If the timestamp is situated in 1970, I would expect the year to be 1970.

Reply:
Yeah, I suppose this is taking `UTC` year instead of `Etc/GMT+2` year. Will fix.

Reply:
It turns out `Etc/GMT+X == UTC-X` so this was actually ok. Weird mapping.